Blog post - PCIe features (#304)
3mdeb/news-and-ideas: blog/content/post/2020-08-05-PCIe-features.md

---
title: What features PCIe has?
abstract: 'Introductory blog post to PCIe features.
          In this article you can read what PCIe capability is
          and see examples of such capabilities.'
cover: /covers/PCIExpress.jpg
author: marek.kasiewicz
layout: post
published: true
date: 2020-08-05
archives: "2020"

tags:
  - Firmware
categories:

>   - Firmware

Reviewer: @Sakoram you should remove the categories that do not fit into this blog post

  - IoT
  - Miscellaneous
  - OS Dev
  - App Dev
  - Security
  - Manufacturing

---

## What this blog is about

> Goal of this post is to introduce readers to the concept of PCIe capabilities.

Reviewer: ```suggestion
Goal of this post is to introduce readers to the concept of PCI Express capabilities.
```

It briefly describes what capability is and how it works. It also lists some

> example capability described in PCIe specification.

Reviewer: ```suggestion
example capability described in PCI Express (PCIe) specification.
```

## Introduction

> Every device in the PCIe tree has at least one function. Each function has its

Reviewer: I would write something more about PCI topology. Jumping straight to the device and functions is not so clear.

> configuration space. Configuration space is device registers mapped to either

Reviewer: What is the configuration space for? What can we do with it? It is okay to list raw facts about PCIe devices, but leaving them without explanation dramatically reduces the readability of the post.

capability in linked list.
- **Linked list of capabilities**  
Each capability has its own header, it contains ID of the
capability, its version and pointer to the next capability in the list. Beneath
this header are mapped registers needed by capabilities.

To check if a given capability is implemented by function, a software has to
search through the list and check if a given capability ID is present in it.

## What PCIe capability actually is

It is nothing more than a predefined feature of the device's function, a feature
that is known to be possible to be implemented in all devices, but most of them
are optional. Every capability has registers in a device mapped to configuration
space. Those registers are a kind of interface to the capability. Some are read
only and contain information needed by the software, others are read/write and
can be used to pass some information to the function.

## List of capabilities structures:
(This list can be incomplete)

- **PCI Express Capability register block**

This is one of the most important capability structures, it must be present in
all PCIe functions. It contains information about:

1. Device e.g.:  

> Maximum supported payload size, maximum accepted latency

Reviewer: Short explanation what payload are we talking about and what the latency refers to

2. Link e.g.:  
Link max speed, link max width, link latency, port number for given link
3. Port e.g.:  
Physical slot number, Hot-Plug capability

**Hot-Plug**

Registers responsible for this capability are located in the Capability register
block. This capability allows change of device in the PCIe slot at runtime.

**Baseline Error Reporting**

> Bits corresponding to this capability are located partly in the configuration

Reviewer: ```suggestion
Bits corresponding to this capability are located partially in the configuration
```

> header, partly in the Capability register block. All of them must be present in

Reviewer: ```suggestion
header, partially in the Capability register block. All of them must be present in
```

all PCIe functions.

- **Power Management**

It permits the OS to manage the power environment of a Function directly by

> accessing registers corresponding to this capability. OS can set device stat to

Reviewer: ```suggestion
accessing registers corresponding to this capability. OS can set device state to
```

one of 4 states. 2 of them are mandatory, and 2 can be optionally implemented.
D0 state is a state where the function is fully operational and uses full power,
every next state (D1,D2,D3) uses lower power, but also takes more time to recover
to D0.

- **Message Signaled Interrupts**

> PCI has pins to let the central interrupt controller know that it needs service.

Reviewer: MSI is a write to memory, it does not use wired signals or pins. This is true in case of legacy INTx

It significantly increased the number of needed lines. In PCIe instead of
additional lines special messages are used to signal interrupt.
These “messages” are actually memory writes to addresses corresponding to a given
function. Address of memory to write and data being written are located in MSI
capability structure.

___

**All following capabilities are optional.**

___

- **EXtended Message Signaled Interrupts**

Extends MSI from 32 possible vectors to 2048 by placing a table containing
addresses and messages of each vector in RAM instead of device registers.

> - **DYNAMIC POWER ALLOCATION**

Reviewer: Why captial letters?

This capability extends power states by additional 32 states between D0 and D1.

- **Power Budgeting**

Its goal is to allocate power for PCIe hot plug devices that are added to the
system during runtime. This ensures that the system can allocate the proper
amount of power and cooling for these devices.

- **Advanced Error Reporting**

It provides additional registers that give error handling software more
information to work with in diagnosing and recovering from problems.

- **Virtual Channel**

It adds buffers that act as queues for outgoing packets. This capability allows
Device to assign one of 8 priorities to packets. Every priority has its own
buffer. Packets from buffers with higher priorities are sent more frequently.

- **TLP Processing Hints**

It adds the possibility to use cache inside switches and the root complex to

> alleviate the need for RAM usage.

Reviewer: Okay, but what is the gain form this capability? Do we get more free RAM?

- **Resizable BAR Capability**

It replaces the BAR registers from HEADER with bigger registers in this
capability structure, allowing bigger space allocation for functions.

- **Access Control Services**

The PCIe specification allows for peer-to-peer transactions.  This means that it
is possible and even desirable in some cases for one PCIe endpoint to send data
directly to another endpoint without having to go through the Root Complex.
ACS provides a mechanism by which a Peer-to-Peer PCIe transaction can be forced
to go up through the PCIe Root Complex. ACS can be thought of as a kind of

> gate-keeper - preventing unauthorized transactions from occurring.

Reviewer: This is explained very well. Try to do something similar in other capabilities as well

- **Multicast**

It enables sending packets to more than one device eliminating the need for the
host to write a unicast packet multiple times to each endpoint which improves
efficiency of the CPU.

- **Alternative Routing ID-Interpretation**

PCIe is p2p communication so there's always one device in every bus. The
Alternative Routing enables to use bits which used to represent device id to
represent function number. It can extend the number of functions from 3 bits to 8.

## Summary

PCIe devices use registers mapped into memory space to allow the OS to know what
device’s feature set is. These registers are grouped into capabilities structures
defined by PCI Special Interest Group in PCIe specification.  
This blog post was only an introduction to PCIe capabilities, the idea is to

> describe some of them in detail in future posts.

Reviewer: Please include our common summary too:
```
If you think we can help in improving the security of your firmware or you
looking for someone who can boost your product by leveraging advanced features
of used hardware platform, feel free to [book a call with us](https://calendly.com/3mdeb/consulting-remote-meeting)
or drop us email to `contact<at>3mdeb<dot>com`. If you are interested in similar
content feel free to [sign up to our newsletter](http://eepurl.com/gfoekD)

```

---
title: What features PCIe has?
abstract: 'Introductory blog post to PCIe features.
          In this article you can read what PCIe capability is
          and see examples of such capabilities.'
cover: /covers/PCIExpress.jpg
author: marek.kasiewicz
layout: post
published: true
date: 2020-08-05
archives: "2020"

tags:
  - Firmware
categories:
  - Firmware

---

## What this blog is about

> Goal of this post is to introduce readers to the concept of PCI Express

Reviewer 2: The goal

> capabilities. It briefly describes what capability is and how it works. It also

Reviewer 2: capabilities capability

lists some example capability described in PCI Express (PCIe) specification.

## Introduction

> PCI Express uses inverted tree topology. It starts from the Root Complex which

Reviewer 2: I miss some one introductory sentence here... I know that you are an engineer and not a writer, but try to add just some opening before starting with your topic

Author: Yes, when I read this post now I can see how far it is from a post written by a good writer ^^
As for the introduction, what would you say about adding something like this:
"
Probably everyone knows about the existence of the PCIe standard. It is commonly
used in PCs to connect various devices to the motherboard. Every possible modern
graphic card uses PCIe, most sound cards use it, it starts to become the standard
even for SSDs, and that's just a tip of the iceberg. But how does the OS communicate
with all of these devices and how does it know what these devices are capable of?
This post will bring you closer to knowing the answer for those questions, but
let's start from the beginning...
"

> is connected directly to the CPU and RAM. Root Complex branches to Switches and

Reviewer 2: Root Complex (...) sentence miss a verb. Should this dot be between those two sentences? Because for me you are listing what is connected to what, so maybe merge it all into one sentence?

Author: Do you mean this sentence: "Root Complex branches to Switches and
Endpoints"?   "branches" is a verb. 

In polish: "kompleks główny rozgałęzia się do przełączników i punktów końcowych"

I'm not sure if it should be one gigantic sentence separated by commas. I think that it works good enough as a multiple shorts sentences, but I can be wrong. Let me know about your final opinion.

> Endpoints. Switches branch founder and can be connected to more Switches or to

Reviewer 2: It switches [missed subject in the sentence]

Author: In the sentence "Switches branch founder and can be connected to more Switches or to Endpoints"  "Switches" is a subject, and "branch" is a verb.

> Endpoints. Endpoints reside at the bottom of branches and have only one

Reviewer 2: too many repetitions: endpoints, switches and so on... maybe again, merge sentences int one bigger, to avoid repetitions?

upstream Port (facing toward the Root). By comparison, a Switch may have several

> downstream Ports, but can only one upstream Port. Possible are also Bridges

Reviewer 2: can only have,  I assume

and Legacy Endpoints. Bridges are interfaces to other buses such as PCI. Legacy
Endpoints are devices which were designed for the operation of an older bus like
PCI‐X but now have a PCIe interface.

![PCIe topology](/img/Example_PCI_Express_Topology.png)  
Every Switch, Bridge and Endpoint is a PCIe Device. All Devices have at least one

> Function, and each Function has its Configuration Space which is Device registers

Reviewer 2: Function must be in capital here? samely Configuration?

mapped to either I/O memory space (used mostly by PCI Devices) or to normal

> memory space. By writing or reading from this memory OS can exchange information

Reviewer 2: memory, OS

> with a Function. For example OS can check the vendor of a Device by checking the

Reviewer 2: check checking  ( try to use [thesaurus](thesaurus.com)) it will ease your writing

Vendor ID field in Configuration Space Header of this device function.

###### In the Configuration Space we can find most importantly 2 structures:

- **Header**  
![config header](/img/Pci-config-space.png)  
It contains information like vendor ID, and Device ID, pointers to memory

> assigned to the Function and pointer to first capability in linked list.

Reviewer 2: too nested line [to...to]

- **Linked list of capabilities**  

> Each capability has its own header, it contains ID of the

Reviewer 2: that instead of it

capability, its version and pointer to the next capability in the list. Beneath
this header are mapped registers needed by the capability.

> To check if a given capability is implemented by Function, a software has to

Reviewer 2: by the Function?

search through the list and check if a given capability ID is present in it.

## What PCIe capability actually is

> It is nothing more than a predefined feature of the Device Function, a feature

Reviewer 2: feature, feature

> that is known to be possible to be implemented in all Function, but most of them

Reviewer 3: Either `all Functions` or `every Function`

> that is known to be possible to be implemented in every Function, but most of them

Reviewer 2: to be, to be

are optional. Every capability has registers in a Device mapped to configuration

> space. Those registers are a kind of interface to the capability. Some are read

Reviewer 2: of an interface

only and contain information needed by the software, others are read/write and
can be used to pass information to the Function.

## List of capabilities structures:
(This list can be incomplete)

- **PCI Express Capability register block**

This is one of the most important capability structure, it must be present in
all PCIe Functions. It is a collection of various information about:

1. **Device** e.g.:  

    - Maximum payload of Transaction Layer Packet size that the Function can support,
    - Maximum accepted change latency between power states.  

2. **Link** e.g.:  

    - Link max speed \[GT/s\]
    - Link max width (number of lines)
    - Link change latency between power states
    - Port number for given Link  

3. **Port** e.g.:
  
    - Physical slot number which is a chassis unique identifier for a slot.  

4. **Hot-Plug**

    Registers responsible for this capability are located in the Capability register
    block. This capability allows change of Device in the PCIe slot at runtime.
    By writing to registers corresponding to this capability OS can let the Root or

>     Switch Ports know to power Off or power On. After power is turned Off an user

Reviewer 2: known

    can safely remove the Device, after inserting new one power should be turned On
    to its Port.

5. **Baseline Error Reporting**

    Bits corresponding to this capability are located partially in the configuration
    header, partially in the Capability register block. All of them must be present in
    all PCIe Functions. Some bits of this capability are used to set error reporting,
    others store status of errors.

- **Power Management**

OS can manage the power environment of a Function directly by accessing registers

> corresponding to this capability. OS can set Device state to one of 4 states. 2

Reviewer 2: of the four

> of them (D0, D3) are mandatory and other 2 (D2, D3) can be optionally implemented.

Reviewer 2: two of them

D0 state is a state where the Function is fully operational and uses full power,

> every next state (D1,D2,D3) uses lower power, but also takes more time to recover

Reviewer 3: Here are D0, D1, D2 and D3, while the list above doesn't have D1 but has D4 instead. One of those is probably wrong

Author: you right. there are 4 states: D0 D1 D2 D3.

to D0.

- **Message Signaled Interrupts**

> PCI has pins to let the central interrupt controller know that it needs service.

Reviewer 2: a service or to be serviced

This improves efficiency of the CPU which does not need to check every Function
periodically, but it also significantly increases the number of needed lines. In
PCIe instead of additional lines special messages are used to signal interrupt.

> Message Signaled Interrupts (MSI) allows the Function to write a small amount of

Reviewer 2: allow

interrupt-describing data to a special memory-mapped I/O address, and the interrupt

> controller then delivers the corresponding interrupt to a processor. Data and

Reviewer 2: , then deliver

Author: Why? It is the third person of the present simple tense, so the "s" suffix should be added to the verb. Isn't it? 

But I agree that this sentence is too long. I'm changing it to:
"Message Signaled Interrupts (MSI) allow the Function to write a small amount of
interrupt-describing data to a special memory-mapped I/O address. The interrupt
controller then delivers the corresponding interrupt to a processor."

> the address to which they are being written are located in MSI capability structure.

Reviewer 2: [just] address

___

**All following capabilities are optional.**

___

> - **EXtended Message Signaled Interrupts**

Reviewer 3: ```suggestion
- **Extended Message Signaled Interrupts**
```

- **Dynamic Power Allocation**

This capability extends power states by additional 32 states between D0 and D1.

- **Power Budgeting**

Its goal is to allocate power for PCIe hot plug Devices that are added to the
system during runtime. This ensures that the system can allocate the proper
amount of power and cooling for these Devices.

- **Advanced Error Reporting**

It provides additional registers that give error handling software more
information to work with in diagnosing and recovering from problems.

- **Virtual Channel**

It adds buffers that act as queues for outgoing packets. This capability allows
Device to assign one of 8 priorities to packets. Every priority has its own
buffer. Packets from buffers with higher priorities are sent more frequently.

- **TLP Processing Hints**

It adds the possibility to use cache inside Switches and the root complex to
alleviate the need for RAM usage. The idea is similar to the idea of ​​adding
cache to processors; adding a small amount of very fast memory to save soon
needed data is a more efficient approach than saving this data in bigger and
slower RAM.

- **Resizable BAR Capability**

> It replaces the BAR registers from HEADER with bigger registers in this

Reviewer 3: Is there a reason why `HEADER` is all caps?

Author: No, I can't recall why I wrote it this way. Maybe, I copied this part from somewhere. I'm changing it.

capability structure, allowing bigger memory space allocation for Functions.

- **Access Control Services**

> The PCIe specification allows for peer-to-peer transactions. This means that it
> is possible and even desirable in some cases for one PCIe Endpoint to send data

Reviewer 3: The second sentence suggests that the fact that it is sometimes desirable is a result of the specification
```suggestion
The PCIe specification allows for peer-to-peer transactions. It
is possible and even desirable in some cases for one PCIe Endpoint to send data
```
